File: backend/app/schemas/execution.py
from datetime import date, datetime
from pydantic import BaseModel


class ExecutionDetailCreate(BaseModel):
    employee_id: str
    work_hours: float = 0
    work_content: str | None = None
    is_leave: bool = False
    leave_reason: str | None = None


class DailyExecutionCreate(BaseModel):
    project_id: str
    record_date: date
    seq_number: int = 1

    # 费用科目
    subcontract_fee: float = 0
    inhouse_personnel: float = 0
    enterprise_personnel: float = 0
    dispatched_personnel: float = 0
    relevant_fee: float = 0
    material_fee: float = 0
    labor_fee: float = 0
    rental_fee: float = 0
    transport_fee: float = 0
    office_fee: float = 0
    entertainment_fee: float = 0
    other_fee: float = 0
    travel_fee: float = 0
    bidding_fee: float = 0
    commission_fee: float = 0
    tax_fee: float = 0

    # 备注
    remark: str | None = None
    registrant: str = ""

    # 人员明细
    details: list[ExecutionDetailCreate] = []


class ExecutionDetailResponse(BaseModel):
    id: str
    employee_id: str
    employee_name: str | None = None
    work_hours: float
    daily_rate: float
    cost: float
    work_content: str | None = None
    is_leave: bool
    leave_reason: str | None = None

    class Config:
        from_attributes = True


class DailyExecutionResponse(BaseModel):
    id: str
    project_id: str
    project_name: str | None = None
    record_date: date
    seq_number: int

    subcontract_fee: float
    inhouse_personnel: float
    enterprise_personnel: float
    dispatched_personnel: float
    relevant_fee: float
    material_fee: float
    labor_fee: float
    rental_fee: float
    transport_fee: float
    office_fee: float
    entertainment_fee: float
    other_fee: float
    travel_fee: float
    bidding_fee: float
    commission_fee: float
    tax_fee: float

    daily_cost: float
    cumulative_cost: float
    cumulative_profit: float
    profit_rate: float

    remark: str | None = None
    status: str
    registrant: str
    reviewer: str | None = None
    details: list[ExecutionDetailResponse] = []
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/api/execution.ts
import { get, post, put, del } from './request'

export interface ExecutionDetailItem {
  employee_id: string; employee_name?: string | null
  work_hours: number; daily_rate: number; cost: number
  work_content: string | null; is_leave: boolean; leave_reason: string | null
}

export interface ExecutionItem {
  id: string; project_id: string; project_name: string | null
  record_date: string; seq_number: number
  subcontract_fee: number; inhouse_personnel: number
  enterprise_personnel: number; dispatched_personnel: number
  relevant_fee: number; material_fee: number; labor_fee: number
  rental_fee: number; transport_fee: number; office_fee: number
  entertainment_fee: number; other_fee: number; travel_fee: number
  bidding_fee: number; commission_fee: number; tax_fee: number
  daily_cost: number; cumulative_cost: number
  cumulative_profit: number; profit_rate: number
  remark: string | null; status: string; registrant: string; reviewer: string | null
  details: ExecutionDetailItem[]; created_at: string
}

export interface BudgetMap {
  [key: string]: number
}

export interface ExecutionListResponse {
  total: number; items: ExecutionItem[]
  accumulated: BudgetMap; budget: BudgetMap
}

export interface PersonnelDailyItem {
  id: string; record_date: string; employee_name: string
  work_type: string; department: string; personnel_type: string
  project_name: string; work_hours: number; work_content: string
  is_leave: boolean; leave_reason: string; cost: number
}

export interface PersonnelDailyResponse {
  total: number; items: PersonnelDailyItem[]
  total_hours: number; total_cost: number; total_days: number
}

export const executionApi = {
  list: (params?: any) => get<ExecutionListResponse>('/executions', params),
  personnelDaily: (params?: any) => get<PersonnelDailyResponse>('/executions/personnel-daily', params),
  get: (id: string) => get<ExecutionItem>(`/executions/${id}`),
  getByDate: (project_id: string, record_date: string) => get<{ found: boolean; item: ExecutionItem | null }>('/executions/by-date', { project_id, record_date }),
  create: (data: any) => post<ExecutionItem>('/executions', data),
  update: (id: string, data: any) => put<ExecutionItem>(`/executions/${id}`, data),
  delete: (id: string) => del<{ message: string }>(`/executions/${id}`),
}
